### DISCUSSION TITLE
Remove next-router

### DISCUSSION THREAD

**Discussion by vincent-thomas** [vercel/next.js#36928] [Ideas]
Can you remove the next-router from the bundle and use a third-party router like `react-router-dom`? That would be a really nice feature. file-routing is pretty much the only thing holding me back from next.js

**Comment by icyJoseph** [comment]
Could you expand a little more on the last bit? In what form is it holding you back?

**Reply by vincent-thomas** [reply]
Sorry. When i write a website, i like components dependencies to be as close as possible (in the file system). For example this:
```
pages/
  home/
    components/...     - components to home route only
    utils/...
    home.route.tsx
    home.module.scss
    home.stories.tsx
    home.spec.tsx
```

Nextjs doesn't allow me to do that. Is there a way to disable file/folder-based routing, and use some sort of manual routing from next (kina like react-router, but ssr)?

**Reply by icyJoseph** [reply]
First of all, I'd say that binding yourself to a file structure, is both kinda good, and kinda bad. I work as a consultant, and I have tested quite a lot of file structures, and at the end of the day, it is really about finding one that makes people more efficient and comfortable. Nothing to test this like a Friday afternoon bug, that must be fixed immediately, then you see how much the file structure is actually helping, versus how pretty, or organised it looks. Though, anchoring to a particular paradigm, won't always work out, and just saying, "but this is what works for me", won't cut it in a team setting.

With that out of the way, I'd say that this is good approach to prevent a big ball of mud. As I see it, home page doesn't care about the existence of other pages; they might as well not exist. Though, the benefit of each component being scoped, might be seen as a limitation by those who want components to be reusable, independently of the call context.

However, there's a few things regarding this structure that might conflict with the overall goal of Next.js (a production framework), for example dynamic paths, rewrites, and nested routes. You'd be also missing on `api/routes`.

I'd build upon your structure by:

```
routes/
  home/
    components/
    utils/
    home.route.tsx
    home.module.scss
    home.stories.tsx
    home.spec.tsx
    home.data.tsx // you can define something like `getServerSideProps` here
pages/
   index.tsx
```

Where `pages/index.tsx`:

```tsx
// using absolute imports
import {Home, homeLayout} from 'routes/home/home.route';
export {getServerSideProps} from 'routes/home/home.data';

Home.getLayout = homeLayout;

export default Home;
```

> Using this https://nextjs.org/docs/basic-features/layouts#per-page-layouts for per page layouts. 

Not saying this is prettier, not better, not more functional, but it'd allow you to keep your structure, without much compromise on the framework. If you ever decide to move away, then the `routes`, except for a bunch of TypeScript, would be reusable.

Something like `react-router-dom` will force you to chose a `path` string, and map it to an element, or component on the `path`, while in this case, we use the file system for that.

This also leaves the door open for `api/routes` which are not React components at all.

One could also flip the coin, and do `components/<route>/` and `utils/<route>/`, but again, this would vary from team to team, and which ever influences they receive.

Here's a write about how Vercel does it, albeit short, https://github.com/vercel/swr/discussions/1408#discussioncomment-1383367, it has good insight, into seeing components as free standing elements, which do not care about the route that uses them. 

Last but not least, from what I understand, if you choose to not use the file system routing, you need to then build your own, through a custom server. `expo` uses the file system routing, but it talks about a way to use `react-navigation` instead, but, I imagine, that's a rabbit hole you don't want to explore. Googling about this yields some information that leads me to belief that, if this is really a deal breaker, and no trade-off satisfies you, then perhaps you shouldn't do it.

**Reply by vincent-thomas** [reply]
Thanks for the detailed answer! I saw this proterty in the next config `useFileSystemPublicRoutes`. This allows me to remove the next/router. What purpose does this serve (or at all) in this context?

**Reply by icyJoseph** [reply]
I am not sure, I've never used it. Hopefully you can find a project online that has used it, or someone who has played with it can help you out. I'll try out myself sometime, but I don't think I can be of any further help here.

**Reply by vincent-thomas** [reply]
Okay, thanks for your time and answer!
